spanner/google/cloud/spanner/streamed.py

```diff
@@ -122,17 +130,50 @@ def _merge_values(self, values):
             field = self.fields[index]
             self._current_row.append(_parse_value_pb(value, field.type))
             if len(self._current_row) == width:
-                self._rows.append(self._current_row)
+                self._pending_rows.append(self._current_row)
                 self._current_row = []
 
+    def _flush_pending_rows(self):
+        """Helper for :meth:`consume_next`."""
+        flushed = self._pending_rows[:]
+        self._pending_rows[:] = ()
+        self._complete_rows.extend(flushed)
+
+    def _do_restart(self):
+        """Helper for :meth:`consume_next`."""
+        self._pending_chunk = None
+        self._pending_rows[:] = ()
+        self._current_row[:] = ()
```

Comment:
It is not safe to reset all these under all conditions. It depends on where you got the resume token. Essentially you need to clear out everything you got after the resume token. But some of this might have come before the resume token. For eg:
Row just has a single string column and these are the partial result sets you get

PRS1: value: "h", chunked: true, resume_token: "zjgfdkjf"
PRS2: value: "e", chunked: true, resume_token: Null

Say when you try to read next PRS3 the stream breaks. So you resume using the token yielded by PRS1. The resumed stream will yield PRS2 onwards. But since you cleared pending_chunk, you lost PRS1.

Reply:
I think this will be easier to implement if you keep on buffering PRS till you see a resume token and only after that merge them/parse them.